src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -0,0 +1,248 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules
+{
+    internal class CollectionRulePipeline : Pipeline
+    {
+        // The executor of the action list for the collection rule.
+        private readonly ActionListExecutor _actionListExecutor;
+
+        // The endpiont that represents the process on which the collection rule is executed.
+        private readonly IEndpointInfo _endpointInfo;
+
+        // The rule description that determines the behavior of the pipeline.
+        private readonly CollectionRuleOptions _ruleOptions;
+        
+        // Task completion source for signalling when the pipeline has finished starting.
+        private readonly TaskCompletionSource<object> _startedSource =
+            new(TaskCreationOptions.RunContinuationsAsynchronously);
+        
+        // Operations for getting triger information.
+        private readonly ICollectionRuleTriggerOperations _triggerOperations;
+
+        public CollectionRulePipeline(
+            ActionListExecutor actionListExecutor,
+            ICollectionRuleTriggerOperations triggerOperations,
+            CollectionRuleOptions ruleOptions,
+            IEndpointInfo endpointInfo)
+        {
+            _actionListExecutor = actionListExecutor ?? throw new ArgumentNullException(nameof(actionListExecutor));
+            _endpointInfo = endpointInfo ?? throw new ArgumentNullException(nameof(endpointInfo));
+            _ruleOptions = ruleOptions ?? throw new ArgumentNullException(nameof(ruleOptions));
+            _triggerOperations = triggerOperations ?? throw new ArgumentNullException(nameof(triggerOperations));
+        }
+
+        /// <summary>
+        /// Starts the execution of the pipeline without waiting for it to run to completion.
+        /// </summary>
+        /// <remarks>
+        /// If the specified trigger is a startup trigger, this method will complete when the
+        /// action list has completed execution. If the specified trigger is not a startup
+        /// trigger, this method will complete after the trigger has been started.
+        /// </remarks>
+        public async Task StartAsync(CancellationToken token)
+        {
+            var runTask = RunAsync(token);
+
+            await _startedSource.WithCancellation(token).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Runs the pipeline to completion.
+        /// </summary>
+        /// <remarks>
+        /// The pipeline will only successfully complete in the following scenarios:
+        /// (1) the trigger is a startup trigger and the action list successfully executes once.
+        /// (2) without a specified action count window duration, the number of action list executions equals the action count limit.
+        /// </remarks>
+        protected override async Task OnRun(CancellationToken token)
```

Comment:
I don't think this is necessary for this PR, but maybe this method needs to be refactored into smaller parts.

Reply:
Yeah, I was thinking of doing that too. I'll think about if I can do it now or in a subsequent change.